File: api/schemas/backtest.py
from pydantic import BaseModel
from typing import Optional, List


class BacktestRequest(BaseModel):
    config_yaml: str
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    initial_cash: Optional[float] = None
    data_path: Optional[str] = None


class BacktestSummary(BaseModel):
    total_return: float = 0
    annual_return: float = 0
    sharpe_ratio: float = 0
    max_drawdown: float = 0
    win_rate: float = 0
    total_trades: int = 0
    final_cash: float = 0
    total_commission: float = 0
    sortino_ratio: float = 0
    calmar_ratio: float = 0
    profit_factor: float = 0
    avg_trade_pnl: float = 0
    trading_days: int = 0


class BacktestResult(BaseModel):
    id: str
    status: str
    summary: BacktestSummary = BacktestSummary()
    config_info: dict = {}
    warnings: List[str] = []
    errors: List[str] = []
    output_files: dict = {}
    created_at: Optional[str] = None
    completed_at: Optional[str] = None


class BacktestTemplate(BaseModel):
    name: str
    description: str
    yaml: str

File: frontend/src/api/backtest.ts
import { get, post } from './index'

export interface BacktestRequest {
  config_yaml: string
  start_date?: string
  end_date?: string
  initial_cash?: number
  data_path?: string
}

export interface BacktestSummary {
  total_return: number
  annual_return: number
  sharpe_ratio: number
  max_drawdown: number
  win_rate: number
  total_trades: number
  final_cash: number
  total_commission: number
  sortino_ratio: number
  calmar_ratio: number
  profit_factor: number
  avg_trade_pnl: number
  trading_days: number
}

export interface BacktestResult {
  id: string
  status: string
  summary: BacktestSummary
  config_info: Record<string, any>
  warnings: string[]
  errors: string[]
  output_files: Record<string, any>
  created_at?: string
  completed_at?: string
}

export interface BacktestTemplate {
  name: string
  description: string
  yaml: string
}

export const backtestApi = {
  run: (data: BacktestRequest) =>
    post<BacktestResult>('/backtest/run', data),

  getResult: (id: string) =>
    get<BacktestResult>(`/backtest/${id}`),

  getHistory: (params?: { limit?: number; offset?: number }) =>
    get<BacktestResult[]>('/backtest/history', { params }),

  getTemplates: () =>
    get<BacktestTemplate[]>('/backtest/templates'),
}
